### ISSUE TITLE
votes,posts,views not pluralized properly

### ISSUE THREAD

**Issue by barisusakli** [NodeBB/NodeBB#6601]

**Comment by Dravere** [comment]
Just in case you don't know it. Pluralization is very simple in English and can be extremely difficult in other languages. If you introduce it, please do it properly and not half-assed as with many native English software solutions.

As a nice summary towards internationalization: https://www.youtube.com/watch?v=0j74jcxSunY (plurals start at [2:46](https://youtu.be/0j74jcxSunY?t=2m46s)).

**Comment by julianlam** [comment]
Yes, this is why the issue has been opened, because we were notified that our "votes/posts/views" is not pluralised at all (though this is mostly a design decision)... but that also, pluralising it would actually involve quite a bit of work since some languages handle it differently :expressionless:

**Comment by Dravere** [comment]
Ah ok, very nice. And good luck :smile:

**Comment by julianlam** [comment]
_Disregarding_ the fact that this means all of our pluralisation efforts are probably wrong, I do want to point out that we have absolutely no idea how to resolve this.

Does anyone speak a language here that has more than singular and plural forms, and can advise us by providing some examples?

We need to turn something like "posts" into a set of translatable **language keys**, which can be sent to Transifex for localisation into their respective languages.

Right now we would just use `[[global:post]]` and `[[global:posts]]`, but we don't know how the other two forms can be used.

**Comment by julianlam** [comment]
That's a simpler example, am ore complicated example would be something like "You have `NUM` unread notifications" (I don't think this is an actual string, but am using it as an example). The entire sentence itself has to be translated as the numeral and form has different placement in different languages, so we'll need that in the other two(?) forms as well.

**Comment by Dravere** [comment]
I would suggest to get some inspiration from GNU [gettext](http://www.gnu.org/software/gettext/manual/gettext.html). That is widely used in the Linux world for i18n. They have chapters about plurals from the [programmers view](http://www.gnu.org/software/gettext/manual/gettext.html#Plural-forms) and [translators view](http://www.gnu.org/software/gettext/manual/gettext.html#Translating-plural-forms). `gettext` uses POT files that seem to be supported by Transifex.

I don't know what is available in Javascript in this regard. Searching for `gettext` returns several results:
https://npms.io/search?q=gettext

There seems to also be Javascript specific solutions not aligned with `gettext` or POT files. Like it seems this library: https://www.i18next.com/overview/introduction

Hope this can get you started.

**Comment by julianlam** [comment]
[Mozilla's Project Fluent](https://projectfluent.org/) seems to be a step in the right direction. What do you think @Dravere?

It might mean we need to migrate all translation efforts away from Transifex, though, which is a tough sell.

**Comment by julianlam** [comment]
With regards to Transifex, there are alternatives that look fairly promising... such as [Pontoon](https://github.com/mozilla/pontoon/), which is actually a node app! :laughing:

**Comment by CodeWitchBella** [comment]
The canonical resource for pluralization rules is Unicode CLDR[1]. There are languages with up to 6 plural forms. My native tongue in particular has three... There are libraries which implement those rules. I personally usually use i18next[2], but there are many others and the choice will probably depend on how the code is written currently, so as to not break anything. There is also browser built-in Intl.PluralRules[3,4], which you can use to get info on which version to use if you want to roll your own solution.


[1]: https://www.unicode.org/cldr/cldr-aux/charts/22/supplemental/language_plural_rules.html
[2]: https://www.i18next.com/
[3]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/PluralRules
[4]: https://caniuse.com/intl-pluralrules

**Comment by julianlam** [comment]
https://help.transifex.com/en/articles/6220806-json-with-icu-plurals

I have not forgotten this issue :eyes:

**Comment by mySebbe** [comment]
Current translation code still performs only key lookup plus positional `%1` replacement, so it has no place to select CLDR plural categories; the catalogs consequently retain many `(s)` strings. Since Transifex supports ICU plurals, would an ICU-compatible value format be accepted while keeping existing string values backward-compatible? A safe first slice could add translator tests for `one/few/many/other` and migrate one non-critical counter before changing the full catalog.
